File: examples/javascript/apisdom-client.js
/**
 * Apisdom JavaScript/Node.js Client
 * Cliente robusto con retry y manejo de errores
 * 
 * Documentación: https://github.com/apisdom/docs
 * Versión: 1.0.0
 */

/**
 * Cliente para las APIs de Apisdom
 */
class ApisdClient {
  /**
   * @param {string} apiKey - Tu API Key de apisdom.com/dashboard
   * @param {string} baseUrl - URL base de la API
   */
  constructor(apiKey, baseUrl = 'https://apisdom.com') {
    this.apiKey = apiKey;
    this.baseUrl = baseUrl;
  }

  /**
   * Request con retry automático y backoff exponencial
   * @private
   */
  async _requestWithRetry(method, endpoint, body = null, maxRetries = 3) {
    const url = `${this.baseUrl}${endpoint}`;

    for (let attempt = 0; attempt < maxRetries; attempt++) {
      try {
        const response = await fetch(url, {
          method,
          headers: {
            'X-API-Key': this.apiKey,
            'Content-Type': 'application/json',
          },
          body: body ? JSON.stringify(body) : undefined,
        });

        // Éxito
        if (response.ok) {
          return response.json();
        }

        // Sin créditos - NO reintentar
        if (response.status === 402) {
          throw new CreditosInsuficientesError(
            'Sin créditos. Recarga en: https://apisdom.com/dashboard'
          );
        }

        // Token inválido - NO reintentar
        if (response.status === 401) {
          throw new TokenInvalidoError('Token inválido o expirado');
        }

        // Rate limit - Esperar y reintentar
        if (response.status === 429) {
          const retryAfter = parseInt(response.headers.get('Retry-After') || '60');
          console.log(`Rate limit. Esperando ${retryAfter}s...`);
          await this._sleep(retryAfter * 1000);
          continue;
        }

        // Error servidor - Retry con backoff
        if (response.status >= 500) {
          const waitTime = Math.pow(2, attempt) * 1000;
          console.log(`Error ${response.status}. Retry en ${waitTime}ms...`);
          await this._sleep(waitTime);
          continue;
        }

        throw new Error(`HTTP ${response.status}`);

      } catch (error) {
        if (error instanceof CreditosInsuficientesError) throw error;
        if (error instanceof TokenInvalidoError) throw error;
        
        if (attempt < maxRetries - 1) {
          const waitTime = Math.pow(2, attempt) * 2000;
          await this._sleep(waitTime);
          continue;
        }
        throw error;
      }
    }

    throw new Error(`Falló después de ${maxRetries} intentos`);
  }

  /**
   * @private
   */
  _sleep(ms) {
    return new Promise(resolve => setTimeout(resolve, ms));
  }

  /**
   * Analiza el sentimiento de un texto
   * @param {string} texto - Texto a analizar (máx 5000 caracteres)
   * @returns {Promise<SentimentResponse>}
   */
  async analizarSentimiento(texto) {
    return this._requestWithRetry(
      'POST',
      '/api/v1/sentiment',
      { text: texto }
    );
  }

  /**
   * Detecta contenido tóxico en un texto
   * @param {string} texto - Texto a moderar (máx 5000 caracteres)
   * @returns {Promise<ModerationResponse>}
   */
  async moderarContenido(texto) {
    return this._requestWithRetry(
      'POST',
      '/api/v1/moderacion',
      { text: texto }
    );
  }

  /**
   * Predice valores futuros de una serie temporal
   * @param {string[]} dates - Fechas en formato 'YYYY-MM-DD' (mín 10)
   * @param {number[]} values - Valores numéricos
   * @param {number} periods - Periodos a predecir (1-365)
   * @returns {Promise<PredictionResponse>}
   */
  async predecirSerie(dates, values, periods = 7) {
    if (dates.length !== values.length) {
      throw new Error('dates y values deben tener la misma longitud');
    }
    if (dates.length < 10) {
      throw new Error('Se necesitan al menos 10 puntos de datos');
    }
    if (periods < 1 || periods > 365) {
      throw new Error('periods debe estar entre 1 y 365');
    }

    return this._requestWithRetry(
      'POST',
      '/api/v1/predictions',
      { dates, values, periods }
    );
  }
}

/**
 * Error: Usuario sin créditos
 */
class CreditosInsuficientesError extends Error {
  constructor(message) {
    super(message);
    this.name = 'CreditosInsuficientesError';
  }
}

/**
 * Error: Token inválido o expirado
 */
class TokenInvalidoError extends Error {
  constructor(message) {
    super(message);
    this.name = 'TokenInvalidoError';
  }
}

// TypeScript types (para referencia)
/**
 * @typedef {Object} SentimentResponse
 * @property {string} text - Texto analizado
 * @property {'positive'|'negative'} sentiment - Sentimiento detectado (modelo binario SST-2)
 * @property {number} score - Confianza (0.0 a 1.0)
 * @property {string|null} warning - Aviso si texto truncado
 */

/**
 * @typedef {Object} ModerationResponse
 * @property {string} text - Texto analizado
 * @property {boolean} is_toxic - true si toxicity_score > 0.7
 * @property {number} toxicity_score - Puntuación (0.0 a 1.0)
 * @property {Object} categories - Desglose por categorías
 */

/**
 * @typedef {Object} PredictionResponse
 * @property {Array<{date: string, value: number}>} predictions - Predicciones
 * @property {number} mape - Error porcentual medio
 * @property {string|null} quality_warning - Aviso si MAPE > 40%
 */

// === EJEMPLOS DE USO ===

// Para Node.js, descomenta esta línea:
// module.exports = { ApisdClient, CreditosInsuficientesError, TokenInvalidoError };

// Ejemplo de uso en navegador o Node.js
async function ejemplos() {
  const API_KEY = 'tu_api_key_aqui';
  const client = new ApisdClient(API_KEY);

  console.log('=' .repeat(50));
  console.log('📚 EJEMPLOS DE APISDOM APIs');
  console.log('='.repeat(50));

  // Ejemplo 1: Análisis de sentimiento
  console.log('\n🎭 SENTIMENT API');
  console.log('-'.repeat(40));
  
  try {
    const sentimiento = await client.analizarSentimiento('¡Me encanta este servicio!');
    console.log(`Texto: ${sentimiento.text}`);
    console.log(`Sentimiento: ${sentimiento.sentiment}`);
    console.log(`Confianza: ${(sentimiento.score * 100).toFixed(0)}%`);
  } catch (error) {
    if (error instanceof CreditosInsuficientesError) {
      console.log('⚠️ Sin créditos. Recarga en apisdom.com/dashboard');
    } else {
      console.error('Error:', error.message);
    }
  }

  // Ejemplo 2: Moderación
  console.log('\n🛡️ MODERATION API');
  console.log('-'.repeat(40));
  
  try {
    const moderacion = await client.moderarContenido('Gracias por tu ayuda!');
    console.log(`Texto: ${moderacion.text}`);
    console.log(`¿Tóxico?: ${moderacion.is_toxic ? 'Sí ❌' : 'No ✅'}`);
    console.log(`Score: ${(moderacion.toxicity_score * 100).toFixed(0)}%`);
  } catch (error) {
    console.error('Error:', error.message);
  }

  // Ejemplo 3: Predicción
  console.log('\n📈 PREDICTION API');
  console.log('-'.repeat(40));
  
  try {
    const fechas = [];
    const valores = [100, 105, 102, 108, 115, 112, 120, 118, 125, 130, 128, 135, 140, 138, 145];
    for (let i = 1; i <= 15; i++) {
      fechas.push(`2024-01-${i.toString().padStart(2, '0')}`);
    }
    
    const prediccion = await client.predecirSerie(fechas, valores, 5);
    console.log(`MAPE: ${(prediccion.mape * 100).toFixed(1)}%`);
    console.log('Predicciones:');
    prediccion.predictions.forEach(p => {
      console.log(`  ${p.date}: ${p.value.toFixed(2)}`);
    });
  } catch (error) {
    console.error('Error:', error.message);
  }

  console.log('\n' + '='.repeat(50));
  console.log('✅ Ejemplos completados');
  console.log('📖 Más info: https://github.com/apisdom/docs');
}

// Descomentar para ejecutar ejemplos:
// ejemplos();

File: examples/python/apisdom_client.py
"""
Apisdom Python Client - Cliente robusto con retry y manejo de errores
Versión: 1.0.0
Documentación: https://github.com/apisdom/docs
"""

import time
from typing import Any

import requests


class ApisdClient:
    """
    Cliente robusto para APIs de Apisdom.
    Maneja: retry, backoff, rate limit, sin créditos.
    
    Ejemplo de uso:
        client = ApisdClient("tu_api_key")
        resultado = client.analizar_sentimiento("¡Excelente producto!")
        print(f"Sentimiento: {resultado['sentiment']} ({resultado['score']:.0%})")
    """
    
    def __init__(self, api_key: str, base_url: str = "https://apisdom.com"):
        """
        Inicializa el cliente.
        
        Args:
            api_key: Tu API Key de apisdom.com/dashboard
            base_url: URL base de la API (no cambiar en producción)
        """
        self.api_key = api_key
        self.base_url = base_url
        self.session = requests.Session()
        self.session.headers.update({
            "X-API-Key": api_key,
            "Content-Type": "application/json"
        })
    
    def _request_with_retry(
        self, 
        method: str, 
        endpoint: str, 
        json_data: dict | None = None,
        max_retries: int = 3
    ) -> dict[str, Any]:
        """
        Request con retry automático y backoff exponencial.
        
        Maneja:
        - 429: Espera según Retry-After o backoff
        - 500/502/503: Retry con backoff
        - 402: NO reintenta (sin créditos)
        - 401: NO reintenta (token inválido)
        """
        url = f"{self.base_url}{endpoint}"
        
        for attempt in range(max_retries):
            try:
                response = self.session.request(method, url, json=json_data, timeout=30)
                
                # Éxito
                if response.status_code == 200:
                    return response.json()
                
                # Sin créditos - NO reintentar
                if response.status_code == 402:
                    raise CreditosInsuficientesError(
                        "Sin créditos. Recarga en: https://apisdom.com/dashboard"
                    )
                
                # Token inválido - NO reintentar
                if response.status_code == 401:
                    raise TokenInvalidoError(
                        "Token inválido o expirado. Renueva en el Dashboard."
                    )
                
                # Rate limit - Esperar y reintentar
                if response.status_code == 429:
                    retry_after = int(response.headers.get("Retry-After", 60))
                    print(f"Rate limit. Esperando {retry_after}s...")
                    time.sleep(retry_after)
                    continue
                
                # Error servidor - Retry con backoff
                if response.status_code >= 500:
                    wait_time = (2 ** attempt) * 1  # 1s, 2s, 4s
                    print(f"Error {response.status_code}. Retry en {wait_time}s...")
                    time.sleep(wait_time)
                    continue
                
                # Otros errores - No reintentar
                response.raise_for_status()
                
            except requests.exceptions.Timeout:
                if attempt < max_retries - 1:
                    wait_time = (2 ** attempt) * 2
                    print(f"Timeout. Retry en {wait_time}s...")
                    time.sleep(wait_time)
                    continue
                raise
        
        raise Exception(f"Falló después de {max_retries} intentos")
    
    def analizar_sentimiento(self, texto: str) -> dict:
        """
        Analiza el sentimiento de un texto.
        
        Args:
            texto: Texto a analizar (máx 5000 caracteres)
        
        Returns:
            dict con:
            - text: Texto analizado
            - sentiment: 'positive' o 'negative' (modelo binario SST-2)
            - score: Confianza (0.0 a 1.0)
        
        Raises:
            CreditosInsuficientesError: Sin créditos disponibles
            TokenInvalidoError: API Key inválida o revocada
        """
        return self._request_with_retry(
            "POST", 
            "/api/v1/sentiment",
            {"text": texto}
        )
    
    def moderar_contenido(self, texto: str) -> dict:
        """
        Detecta contenido tóxico en un texto.
        
        Args:
            texto: Texto a moderar (máx 5000 caracteres)
        
        Returns:
            dict con:
            - text: Texto analizado
            - is_toxic: True si toxicity_score > 0.7
            - toxicity_score: Puntuación de toxicidad (0.0 a 1.0)
            - categories: Desglose por categorías
        
        Raises:
            CreditosInsuficientesError: Sin créditos disponibles
            TokenInvalidoError: API Key inválida o revocada
        """
        return self._request_with_retry(
            "POST",
            "/api/v1/moderacion", 
            {"text": texto}
        )
    
    def predecir_serie(
        self, 
        dates: list[str], 
        values: list[float], 
        periods: int = 7
    ) -> dict:
        """
        Predice valores futuros de una serie temporal.
        
        Args:
            dates: Lista de fechas en formato 'YYYY-MM-DD' (mín 10)
            values: Lista de valores numéricos (misma longitud que dates)
            periods: Periodos futuros a predecir (1-365, default 7)
        
        Returns:
            dict con:
            - predictions: Lista de {date, value, lower, upper}
            - mape: Error porcentual medio (0.0 a 1.0)
            - quality_warning: Aviso si MAPE > 40%
        
        Raises:
            CreditosInsuficientesError: Sin créditos disponibles
            TokenInvalidoError: API Key inválida o revocada
            ValueError: Si datos no cumplen requisitos
        """
        if len(dates) != len(values):
            raise ValueError("dates y values deben tener la misma longitud")
        if len(dates) < 10:
            raise ValueError("Se necesitan al menos 10 puntos de datos")
        if not 1 <= periods <= 365:
            raise ValueError("periods debe estar entre 1 y 365")
            
        return self._request_with_retry(
            "POST",
            "/api/v1/predictions",
            {"dates": dates, "values": values, "periods": periods}
        )


class CreditosInsuficientesError(Exception):
    """Usuario sin créditos - mostrar mensaje en UI"""
    pass


class TokenInvalidoError(Exception):
    """Token expirado o inválido - redirigir a login"""
    pass


# === EJEMPLO DE USO ===
if __name__ == "__main__":
    # Reemplaza con tu API Key real
    API_KEY = "tu_api_key_aqui"
    
    client = ApisdClient(API_KEY)
    
    # Ejemplo 1: Análisis de sentimiento
    print("=" * 50)
    print("🎭 ANÁLISIS DE SENTIMIENTO")
    print("=" * 50)
    
    try:
        resultado = client.analizar_sentimiento("¡Me encanta este servicio!")
        print(f"Texto: {resultado['text']}")
        print(f"Sentimiento: {resultado['sentiment']}")
        print(f"Confianza: {resultado['score']:.0%}")
        
    except CreditosInsuficientesError:
        print("⚠️ Sin créditos. Recarga en apisdom.com/dashboard")
    except TokenInvalidoError:
        print("🔒 API Key inválida. Obtén una nueva en el Dashboard")
    
    # Ejemplo 2: Moderación de contenido
    print("\n" + "=" * 50)
    print("🛡️ MODERACIÓN DE CONTENIDO")
    print("=" * 50)
    
    try:
        resultado = client.moderar_contenido("Gracias por tu ayuda, eres genial!")
        print(f"Texto: {resultado['text']}")
        print(f"¿Es tóxico?: {'Sí' if resultado['is_toxic'] else 'No'}")
        print(f"Score: {resultado['toxicity_score']:.0%}")
        
    except CreditosInsuficientesError:
        print("⚠️ Sin créditos. Recarga en apisdom.com/dashboard")
    
    # Ejemplo 3: Predicción de series temporales
    print("\n" + "=" * 50)
    print("📈 PREDICCIÓN DE SERIES TEMPORALES")
    print("=" * 50)
    
    try:
        # Datos de ejemplo: 15 días de ventas
        fechas = [f"2024-01-{i:02d}" for i in range(1, 16)]
        valores = [100, 105, 102, 108, 115, 112, 120, 118, 125, 130, 128, 135, 140, 138, 145]
        
        resultado = client.predecir_serie(fechas, valores, periods=5)
        
        print(f"MAPE (error): {resultado['mape']:.1%}")
        print("\nPredicciones:")
        for pred in resultado['predictions']:
            print(f"  {pred['date']}: {pred['value']:.2f}")
        
        if resultado.get('quality_warning'):
            print(f"\n⚠️ Aviso: {resultado['quality_warning']}")
            
    except CreditosInsuficientesError:
        print("⚠️ Sin créditos. Recarga en apisdom.com/dashboard")
    except ValueError as e:
        print(f"❌ Error en datos: {e}")
